### ISSUE TITLE
Keras2ONNX converts models, but upon importing into Unity, output shape is (-1,-1,-1,-1)

### ISSUE THREAD

**Issue by hellovivian** [Unity-Technologies/barracuda-release#65]
I have written an autoencoder, a GAN, and a GAN trained on the latent space of the autoencoder in Keras. While Keras2ONNX converts my Keras model without complaints (ie. produces an ONNX model), when I import it into Unity, I get many errors that render each model unusable. 

![onnxerros](https://user-images.githubusercontent.com/15935546/85096192-827be080-b1a8-11ea-90e2-b485ab7884da.png)

Namely, the shape outputs of each one of these models (Keras, python3) is (-1,-1,-1,-1). Many layers are swapped out for Identity layers, and so on.
I can think of reasons why the conversion and importing might fail.

1) Autoencoder -- relies on a custom loss and has Conv1D nodes
2) GAN -- discriminator utilizes Conv1D nodes
3) GAN on latent space -- this one is a wasserstein GAN which hacked a Keras class so that it could use gradient penalty (I understand that this one was a long shot, even though the architecture of the generator and discriminator are simple) 

Are these assumptions right? 

And if so... does anyone have any ideas for how I could adapt my models so that they can import into Unity?

**Comment by AlexRibard** [comment]
Hi @hellovivian !
Would it be possible for you to share your ONNX models so I can get a better picture?

* Output shape being (-1,-1,-1,-1) means we were not able to figure out the output shape at compile time.
In itself it is not necessarily a problem.
* If a layer is swapped by `Identity` it probably means that we were not able to load it.
Either because we don't support the layer outright, or it might be some problematic layer option
You can check the supported layers and options here:
https://github.com/Unity-Technologies/barracuda-release/blob/release/1.0.0/Documentation%7E/SupportedOperators.md

It is hard to tell from the screenshot, but it looks like we are complaining about --inputs-as-nchw
If you are exporting from TF, please use this option to tell the converter you want the inputs to be treated as NCHW (channel first), i.e ONNX's default data layout

**Comment by hellovivian** [comment]
Hi @AlexRibard ! Thanks so much for getting back to me, excited to hear from you! :)

Oh! I will have to check if I used that --inputs-as-nchw flag you mentioned. I will get back to you on that as soon as I redo the conversion today or tomorrow.

I can definitely share my ONNX model with you (as well as Keras code / architecture if you need that for reference) through google drive, but how might I go about getting that to you privately?

**Comment by AlexRibard** [comment]
For sure.
You can reach me personally via email: barracuda-support@unity3d.com
The best would be to share:
`.h5` file (keras model) and/or corresponding `.onnx` file. You can initialize the weights randomly, it's ok.

**Comment by Rosthouse** [comment]
I seem to be experiencing the same error when trying to import the [TinyYOLO V2 model ](https://github.com/onnx/models/tree/master/vision/object_detection_segmentation/tiny-yolov2)from the onnx model zoo. Regardless of the opset, the outputs is always as follows:

![TinyYOLO V2 import](https://user-images.githubusercontent.com/3259696/86240245-20a47900-bba1-11ea-923d-5a50b19e5c8c.png)

I tried to regenerate the onnx model as described [here](https://github.com/onnx/onnxmltools#coreml-to-onnx-conversion) using the same CoreML [model as in the onnx zoo](https://github.com/hollance/YOLO-CoreML-MPSNNGraph/blob/master/TinyYOLO-CoreML/TinyYOLO-CoreML/TinyYOLO.mlmodel), but I had the same result.

To clarify: I have been following the turorial by [Andrei Ashikhmin](https://classifai.net/blog/tensorflow-onnx-unity/) which directed me to that model. He used Barracuda 0.4.0-preview as far as I can tell.

**Comment by AlexRibard** [comment]
This should be fine, the outputs being (-1,-1,-1,-1) just means we were not able to infer it's shape at compile time.
But it doesn't impact the execution.
Let me know if your model runs correctly.
Can you share your onnx model, I'll check why the output is that way.

**Comment by JackWei0310** [comment]
I meet the same problem and I do not know how to solve it.
For picture, the .onnx file should have 2 inputs (N,225,225,4) and (N,14,14,1024). The output should be (N,225,225,1). But it mistakes in Unity.
I check each layer is supported by Barracuda, I use keras2onnx to turn .h5 file to .onnx file.
The model is U-net as a generator in Gan.
![2020-07-07](https://user-images.githubusercontent.com/67232658/86766656-c09b5000-c042-11ea-9c88-98f0acbf8512.png)

**Comment by AlexRibard** [comment]
@JackWei0310 it shouldn't be a problem to run your model.
We write out (-1,-1,-1,-1) if we cannot figure out all the output dimensions at compile time.
Given that you have a unknown batch, my guess is that we write out (-1,-1,-1,-1).
I can see how that could be confusing, but it will not stop correctly executing your model.
I will update the inspector accordingly.

**Comment by JackWei0310** [comment]
> @JackWei0310 it shouldn't be a problem to run your model.
> We write out (-1,-1,-1,-1) if we cannot figure out all the output dimensions at compile time.
> Given that you have a unknown batch, my guess is that we write out (-1,-1,-1,-1).
> I can see how that could be confusing, but it will not stop correctly executing your model.
> I will update the inspector accordingly.

Thanks for your explanation, I still have a question, why my input should be (0,225,225,4) and it showed (0,225,4,225) in Unity?

**Comment by AlexRibard** [comment]
Pytroch (and thus onnx) data layout is NCHW so in your case [0,225,225,4] so 256 channels, 225 height and 4 width
Barracuda is NHWC so in your case this becomes [0,225,4,225]
https://github.com/Unity-Technologies/barracuda-release/blob/release/1.0.0/Documentation%7E/TensorHandling.md

You can confirm, but I would imagine that you have 4 channels and 225x225 height x width
Did you export with the inputs_as_nchw flag with keras2onnx?
https://github.com/Unity-Technologies/barracuda-release/blob/release/1.0.0/Documentation%7E/Exporting.md

**Comment by JackWei0310** [comment]
> Pytroch (and thus onnx) data layout is NCHW so in your case [0,225,225,4] so 256 channels, 225 height and 4 width
> Barracuda is NHWC so in your case this becomes [0,225,4,225]
> https://github.com/Unity-Technologies/barracuda-release/blob/release/1.0.0/Documentation%7E/TensorHandling.md
> 
> You can confirm, but I would imagine that you have 4 channels and 225x225 height x width
> Did you export with the inputs_as_nchw flag with keras2onnx?
> https://github.com/Unity-Technologies/barracuda-release/blob/release/1.0.0/Documentation%7E/Exporting.md

I used keras model and I followed with Unity Document steps to transform:
Here is my codes:
model= load_model('C:/Users/43114/Desktop/model/terrain_generator10000.h5')
onnx_model = keras2onnx.convert_keras(model,         # keras model
                         name="watershed_generator10000",           # the converted ONNX model internal name                     
                         target_opset=9,           # the ONNX version to export the model to
                         channel_first_inputs=None # which inputs to transpose from NHWC to NCHW
                         )

onnx.save_model(onnx_model, "C:/Users/43114/Desktop/model_onnx/watershed_generator10000_v5_v3.onnx")

So where I made mistakes?

**Comment by AlexRibard** [comment]
channel_first_inputs=None -> you should specify your inputs name here as
example channel_first_inputs=["my_input"]

**Comment by FlorentGuinier** [comment]
@JackWei0310 Closing the issue as problem seems to be solved/explained, please reopen if you are still experiencing an issue.
